cryodrgn/dataset.py

```diff
@@ -1,397 +1,112 @@
-
-    def __init__(self, mrcfile, norm=None, ind=None, lazy=False, use_cupy=False):
-        self.use_cupy = use_cupy
-        particles = load_particles(mrcfile, lazy=lazy)
-        self.lazy = lazy
-        if ind is not None:
-            particles = particles[ind]
+        self.N = self.src.n
+        self.D = ny
+        self.invert_data = invert_data
+        self.window = window_mask(ny - 1, window_r, 0.99) if window else None
+        self.max_threads = min(max_threads, mp.cpu_count())
+        self.norm = norm or self.estimate_normalization(n=1000)
+        self.device = device
 
-        self.particles = particles
-        self.N = len(particles)
-        if self.lazy:
-            self.D = particles[0].get().shape[0]  # ny + 1 after symmetrizing HT
-        else:
-            self.D = particles.shape[1]  # ny + 1 after symmetrizing HT
+    def estimate_normalization(self, n=None):
+        n = min(n, self.N) if n is not None else self.N
+        indices = range(0, self.N, self.N // n)
+        imgs = self.src.images(indices)
 
-        logger.info(f"Loaded {len(particles)} {self.D}x{self.D} images")
-        if norm is None:
-            norm = list(self.calc_statistic())
-            norm[0] = 0
+        if self.window is not None:
+            imgs[..., : self.D - 1, : self.D - 1] *= self.window
 
-        if not lazy:
-            self.particles = (self.particles - norm[0]) / norm[1]
+        fft.ht2_center(
+            imgs[..., :-1, :-1],
+            inplace=True,
+            chunksize=1000,
+            n_workers=self.max_threads,
+        )
 
-        logger.info("Normalized HT by {} +/- {}".format(*norm))
-        self.norm = norm
+        if self.invert_data:
+            imgs *= -1
 
-    def calc_statistic(self):
-        pp = cp if (self.use_cupy and cp is not None) else np
+        imgs = fft.symmetrize_ht(imgs, preallocated=True)
+        norm = (0, torch.std(imgs))
+        logger.info("Normalizing HT by {} +/- {}".format(*norm))
+        return norm
 
-        if self.lazy:
-            max_size = min(10000, self.N)
-            sample_index = pp.sort(
-                pp.random.choice(
-                    pp.arange(self.N), max(int(0.1 * self.N), max_size), replace=False
-                )
-            )
-            print("--lazy mode, sample 10% of samples to calculate standard error...")
-            data = []
-            for d in sample_index:
-                data.append(self.particles[d].get())
-            data = pp.stack(data, 0)
-            mean, std = pp.mean(data), pp.std(data)
-        else:
-            mean, std = pp.mean(self.particles), pp.std(self.particles)
-        # print(f"std={std}, mean={mean}")
-        return mean, std
+    def _process(self, data):
+        if data.ndim == 2:
+            data = data[np.newaxis, ...]
+        if self.window is not None:
+            data[..., : self.D - 1, : self.D - 1] *= self.window
+        fft.ht2_center(data[..., :-1, :-1], inplace=True)
```

Comment:
If you do the HT directly into the data buffer, then when you come around on the next epoch aren't you going to HT the already-HT'd data?

Reply:
I've gone ahead and implemented this in commit 7f3b87558d77091d39877540c5a6906de0ea41cc

Reply:
From discussions in another conversation, looks like we don't want to go here:
```
Proceed to refactor the behavior of ImageSource in the eager case as to not make copies.
```

So I've reverted that logic now in 194f845863403eac824355ad33b11f6eff737946. All invocations of `images()` (whether lazy or not) now return a copy of the data.